### PR TITLE
Lua interface rework

### LINKED ISSUE #421: Add LUA save/load
Out from #36:

Saving and loading of lua states should be possible.

Research shows, that automatic save/load is really problematic. So I'd add Save/Load callins to lua and make the script save its own state as required. Instead of OnStart, OnContinue can be called or a param be added to OnStart.

Details: Pass lua an object to Save/Load that can be used to Write/Read integers and strings. This is serialized via serializer and appended to the savegame. Should be pretty easy and keeps the savegame small (LUA VM state might be rather large...)

### PR DESCRIPTION
**Pull request by Flamefire**
This switches the lua interface to a C++ wrapper based OO-Design

It also adds a simple test case that runs all Lua functions and does also some minor error checking (I'd love to have unit tests...)

It changes the current interface completely so current lua scripts won't work anymore.
However the transition should be straight forward, see https://github.com/Flamefire/s25client/blob/lua/tests/maps/LuaFunctions.lua

@Spikeone: Have a look at the Lua file: https://github.com/Flamefire/s25client/blob/lua/tests/maps/LuaFunctions.lua and tell me if there is anything to improve for you

Will close #421, #422

Todo: 
- [x] test player join event (I think one cannot set nation for open spots...)
- [x] Run onSettingsReady only on host(?) (setting addons etc should only be done by the host)

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -25,3 +25,6 @@
 [submodule "RTTR/languages"]
 	url = https://github.com/Return-To-The-Roots/languages.git
 	path = RTTR/languages
+[submodule "contrib/kaguya"]
+	path = contrib/kaguya
+	url = https://github.com/satoren/kaguya.git
```
Comment (Flow86) [inline .gitmodules]: we should fork the kaguya repo to the rttr scope if we use it at the end

Comment (Flamefire) [inline .gitmodules]: I thought about that. But I did not see an advantage (for now). The guy maintaining it is pretty active and this way we are better linked to the upstream. We maintain the other repos, hence they are in our scope, but I don't want/need us to maintain kaguya ;)

**Comment by Flamefire** [comment]
In the works: Change game settings via Lua

File: .gitmodules (hunk 1 above)
Comment (Flow86) [inline .gitmodules]: yes but I dont like the idea depending on "some stuff somewhere on the internet" ;-)

**Comment by Flow86** [comment]
one side note: I've authorized travis and appveyor now, appveyor build is running ;-)

File: .gitmodules (hunk 1 above)
Comment (Flamefire) [inline .gitmodules]: We don't really depend on it (in a bad way). This is what submodules are for: Pull in some *specific* source code. So we just pull in another library at a version we choose. Why should we have extra work for maintaining our copy of that repo?

**Comment by Flamefire** [comment]
Ok nice :) Can I have the URL to the badge or can you add the 2 badges (travis, appveyor for master branch) to the readme? Always nice to see a green icon ;)

**Comment by Flamefire** [comment]
@Spikeone It is now possible to change game settings including addons while even allowing the user to change some addons. Have a look at the now extended example script.

Todo: 
- [ ] add player join event (I think one cannot set nation for open spots...)
- [ ] Run onSettingsReady only on host (setting addons etc should only be done by the host)

**Comment by Spikeone** [comment]
Just <3

**Comment by Flamefire** [comment]
Almost complete now. Still need to test the player joining and do some refactoring for names of commands (many are called toggle but are actually set commands)

**Comment by Spikeone** [comment]
So I can close playerslots and add AI players then? Anyway, great work at all :)

**Comment by Flamefire** [comment]
Yes :)

**Comment by Spikeone** [comment]
Uh uh, now we need some input boxes like vbs does, simple yes no selection, pause game, simple input box so we can actually use puzzles or riddles where you have to find like a password or solve something and then... uhhh those possibilitys. Maybe I wont convert my old scripts but rewrite them from scratch

**Comment by Flamefire** [comment]
Currently adding message boxes. But I would not add decisions in the near future as this might cause asyncs...

**Comment by Spikeone** [comment]
Oh and can you also add like pictures in messages somehow? so we can specify any image by x and y offset from the upper left of that message box to be displayed? That might not be a very handy way but... at least it works and one could do nearly everything with that (like tutorial pictures etc.)

**Comment by Flamefire** [comment]
Ok :)

**Comment by Spikeone** [comment]
<3 love it

**Comment by Flamefire** [comment]
Tested and I'd considers this ready to go. Wiki is already updated

**Pull request merged**